Description: Fix python tests for Python 2.7
 python/pythontest2.py: With Python 2.7, gc.collect() doesn't always
 collect all the garbage, so repeat until no further objects are
 freed.  (Closes: #625679)
Author: Olly Betts <olly@survex.com>
Bug-Debian: http://bugs.debian.org/625679
Origin: upstream
Last-Update: 2011-05-17

--- xapian-bindings-1.2.5.orig/python/pythontest2.py
+++ xapian-bindings-1.2.5/python/pythontest2.py
@@ -29,6 +29,17 @@ import xapian
 
 from testsuite import *
 
+def gc_object_count():
+    # Python 2.7 doesn't seem to free all objects even for a full collection,
+    # so collect repeatedly until no further objects get freed.
+    old_count, count = len(gc.get_objects()), 0
+    while True:
+        gc.collect()
+        count = len(gc.get_objects())
+        if count == old_count:
+            return count
+        old_count = count
+
 def setup_database():
     """Set up and return an inmemory database with 5 documents.
 
@@ -222,8 +233,7 @@ def test_matchingterms_iter():
 
     # Check for memory leaks: Prior to 1.2.4 Enquire.matching_terms()
     # leaked references to its members.
-    gc.collect()
-    object_count = len(gc.get_objects())
+    object_count = gc_object_count()
 
     enquire = xapian.Enquire(db)
     enquire.set_query(query)
@@ -239,8 +249,7 @@ def test_matchingterms_iter():
     expect(mterms, ['it', 'two', 'warm', 'was'])
 
     del mterms, mterms2, term, item, enquire, mset
-    gc.collect()
-    expect(object_count, len(gc.get_objects()))
+    expect(object_count, gc_object_count())
 
 def test_queryterms_iter():
     """Test Query term iterator.
@@ -1531,13 +1540,10 @@ def test_leak_mset_items():
     enq.set_query(xapian.Query('drip'))
     mset = enq.get_mset(0, 10)
 
-    gc.collect()
-    object_count = len(gc.get_objects())
+    object_count = gc_object_count()
     # Prior to 1.2.4 this next line leaked an object.
     mset.items
-    gc.collect()
-
-    expect(object_count, len(gc.get_objects()))
+    expect(object_count, gc_object_count())
 
 def test_custom_matchspy():
     class MSpy(xapian.MatchSpy):
